analysis/diabetes_analysis.ipynb

```diff
@@ -223,8 +226,337 @@
    ],
    "source": [
     "# load data\n",
-    "df = pd.read_csv('../data/diabetes.csv')\n",
-    "df.head()"
+    "df_original = pd.read_csv('../data/diabetes.csv')\n",
+    "df_original.head()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "7749a78a-528f-453d-ab13-6f6df87a6aaf",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "# Configure logging\n",
+    "logging.basicConfig(\n",
+    "    filename=\"validation_errors.log\",\n",
+    "    filemode=\"w\",\n",
+    "    format=\"%(asctime)s - %(message)s\",\n",
+    "    level=logging.INFO,\n",
+    ")\n",
+    "\n",
+    "# Define schema\n",
+    "schema = pa.DataFrameSchema(\n",
+    "    { \n",
+    "        \"Outcome\": pa.Column(int, pa.Check.isin([0, 1])),\n",
+    "        \"Pregnancies\": pa.Column(int, pa.Check.between(0, 15), nullable=True),\n",
+    "        \"Glucose\": pa.Column(int, pa.Check.between(0, 230), nullable=True),\n",
+    "        \"BloodPressure\": pa.Column(int, pa.Check.between(0, 140), nullable=True),\n",
+    "        \"SkinThickness\": pa.Column(int, pa.Check.between(0, 80), nullable=True),\n",
+    "        \"Insulin\": pa.Column(int, pa.Check.between(0, 800), nullable=True),\n",
+    "        \"BMI\": pa.Column(float, pa.Check.between(0, 65), nullable=True),\n",
+    "        \"DiabetesPedigreeFunction\": pa.Column(float, pa.Check.between(0, 3), nullable=True),\n",
+    "        \"Age\": pa.Column(int, pa.Check.between(0, 90), nullable=True),\n",
+    "    },\n",
+    "    checks=[\n",
+    "        pa.Check(lambda df: ~df.duplicated().any(), error=\"Duplicate rows found.\"),\n",
+    "        pa.Check(lambda df: ~(df.isna().all(axis=1)).any(), error=\"Empty rows found.\")\n",
+    "    ],\n",
+    "    drop_invalid_rows=False,  # Ensure this is properly closed\n",
+    ")\n",
+    "\n",
+    "# Initialize error cases DataFrame\n",
+    "error_cases = pd.DataFrame()\n",
+    "data = df_original.copy()\n",
+    "\n",
+    "# Validate data and handle errors\n",
+    "try:\n",
+    "    validated_data = schema.validate(data, lazy=True)\n",
+    "except pa.errors.SchemaErrors as e:\n",
+    "    error_cases = e.failure_cases\n",
+    "\n",
+    "    # Convert the error message to a JSON string\n",
+    "    error_message = json.dumps(e.message, indent=2)\n",
+    "    logging.error(\"\\n\" + error_message)\n",
+    "\n",
+    "# Filter out invalid rows based on the error cases\n",
+    "if not error_cases.empty:\n",
+    "    invalid_indices = error_cases[\"index\"].dropna().unique()\n",
+    "    df = (\n",
+    "        data.drop(index=invalid_indices)\n",
+    "        .reset_index(drop=True)\n",
+    "        .drop_duplicates()\n",
+    "        .dropna(how=\"all\")\n",
+    "    )\n",
+    "else:\n",
+    "    df = data\n"
```

Comment:
On top of this, should we save this validated data to a `.csv` in the data folder.

I see in Tiff's Milestone 2 example, the data folder is now structured into `raw` vs. `processed` and saved with processed data. We can use a similar structure and save the validated_data using commands like `df.to_csv("../data/processed/diabetes_validated.csv")`